llvm/labs/lab2/safarov_nurlan/replaceMulWithBitShift.cpp

```diff
@@ -0,0 +1,84 @@
+#include "llvm/IR/Function.h"
+#include "llvm/Pass.h"
+#include "llvm/Passes/PassBuilder.h"
+#include "llvm/Passes/PassPlugin.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/raw_ostream.h"
+#include "llvm/Transforms/Utils/BasicBlockUtils.h"
+#include <vector>
+
+using namespace llvm;
+
+namespace {
+
+bool runPlugin(Function &F) {
+  std::vector<Instruction *> instrDeleteList;
+  for (llvm::BasicBlock &funcBlock : F) {
+    for (llvm::Instruction &instr : funcBlock) {
+      if (llvm::BinaryOperator *binOper = dyn_cast<BinaryOperator>(&instr)) {
+        if (binOper->getOpcode() == llvm::Instruction::BinaryOps::Mul) {
+          Constant *constant = dyn_cast<Constant>(binOper->getOperand(0));
+          int notConstantOperandNumber;
+          if (constant) {
+            notConstantOperandNumber = 1;
+          } else {
+            constant = dyn_cast<Constant>(binOper->getOperand(1));
+            notConstantOperandNumber = 0;
+          }
+          if (constant) {
+            llvm::APInt value = constant->getUniqueInteger();
+            if ((value & (value - 1)) == 0) {
```

Comment:
There is a special method available for checking that Value is a power of 2.

Reply:
Changed.